src/test/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImplIT.java

```diff
@@ -0,0 +1,401 @@
+          );
+    }
+
+    @CartesianTest
+    @CartesianTest.MethodFactory("badData")
+    void shouldThrowBadRequestIfSchemaIsNotCompatibleWithData(
+        SchemaManager.SchemaFormat keyFormat, Object badData
+    ) {
+      var topic = randomTopicName();
+      createTopic(topic);
+      var keySchema = createSchema(
+          "%s-key".formatted(topic),
+          keyFormat.schemaProvider().schemaType(),
+          getProductSchema(keyFormat)
+      );
+
+      produceRecordThen(
+          null, topic, badData, keySchema.getVersion(), null, null)
+          .statusCode(400)
+          .body("message", containsString("Failed to parse data"));
+
+      var valueSchema = createSchema(
+          "%s-value".formatted(topic),
+          keyFormat.schemaProvider().schemaType(),
+          getProductSchema(keyFormat)
+      );
+
+      produceRecordThen(
+          null, topic, null, null, badData, valueSchema.getVersion())
+          .statusCode(400)
+          .body("message", containsString("Failed to parse data"));
+    }
+
+    private static Stream<Arguments> shouldThrowBadRequestForInvalidProtobufData() {
+      return Stream.of(
+          Arguments.of(
+              Map.of("id", "invalid", "name", "hello", "price", 123.45)),
+          Arguments.of(
+              Map.of("id", 10, "name", List.of("hello", "world"), "price", 123.45))
+      );
+    }
+
+    // Special treatment for protobuf data since it's more liberal in schema validation
+    // and there is no way to declare required fields in protobuf.
+    @ParameterizedTest
+    @MethodSource("shouldThrowBadRequestForInvalidProtobufData")
+    void shouldThrowBadRequestForInvalidProtobufData(Object badData) {
+      shouldThrowBadRequestIfSchemaIsNotCompatibleWithData(
+          SchemaManager.SchemaFormat.PROTOBUF, badData
+      );
+    }
+  }
+
+  // Happy path
+  @Nested
+  @TestProfile(NoAccessFilterProfile.class)
+  class HappyPath {
+    private static RecordData jsonData(Object data) {
+      return new RecordData(null, null, data);
+    }
+
+    private final static String PRODUCT_DATA = """
+      {
+        "id": 123,
+        "name": "hello",
+        "price": 123.45,
+        "tags": ["hello", "world"]
+      }
+      """;
+
+    private final static List<RecordData> RECORD_DATA_VALUES = List.of(
+        jsonData(null),
+        jsonData("hello"),
+        jsonData(123),
+        jsonData(123.45),
+        jsonData(true),
+        jsonData(List.of("hello", "world")),
+        jsonData(Collections.singletonMap("hello", "world")),
+        schemaData(SchemaManager.SchemaFormat.JSON, PRODUCT_DATA),
+        schemaData(SchemaManager.SchemaFormat.PROTOBUF, PRODUCT_DATA),
+        schemaData(SchemaManager.SchemaFormat.AVRO, PRODUCT_DATA)
+    );
+
+    private static ArgumentSets testProduceAndConsumeData() {
+      return ArgumentSets
+          .argumentsForFirstParameter(RECORD_DATA_VALUES)
+          .argumentsForNextParameter(RECORD_DATA_VALUES);
+    }
```

Comment:
Same comment as above. The `test` prefix here makes it seem like a test method. 
```suggestion
    /**
     * Valid keys and values inputs used for producing and consuming data.
     * @return the sets of keys and sets of values
     */
    static ArgumentSets validKeysAndValues() {
      return ArgumentSets
          .argumentsForFirstParameter(RECORD_DATA_VALUES)
          .argumentsForNextParameter(RECORD_DATA_VALUES);
    }
```